### PR TITLE
[PR1] Dev-25: Adding Three JS and connection with backend HIL

### PR DESCRIPTION
**Pull request by lipezaballa**
This PR uses Three js to represent the vehicle in the Testing page, it also connects to the HIL backend and allows to send commands and represent the vehicle status according to the information coming from the backend. Play and stop buttons have been implemented to start and stop the simulation with the HIL.
It makes use of useSpring for the animation of the vehicle. We have stopped using the form for the moment to create several buttons that customize default commands.

### CODE REVIEW

File: src/components/InstructionButton/InstructionButton.tsx [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    );
+}
+
+function sendOrder(
+    isOn: boolean,
+    id: number,
+    sendJsonMessage: SendJsonMessage
+) {
+    const controlOrder: ControlOrder = { id: id, state: isOn };
+    console.log(controlOrder);
```
Comment (smorsua) [inline src/components/InstructionButton/InstructionButton.tsx]: Remove console logs

File: src/components/PlayButton/PlayButton.tsx [hunk 2]
```diff
@@ -28,29 +38,75 @@ const buttonVariants = {
...
+            break;
+        case "stop":
+            sendMsgSimultation(sendMessage!, FINISH_SIMULTATION);
+            changeState({ play: state, stop: !state });
+            break;
+    }
+}
+
+function sendMsgSimultation(sendMessage: SendMessage, msg: string) {
+    console.log("Sent: ", msg);
```
Comment (smorsua) [inline src/components/PlayButton/PlayButton.tsx]: Remove log

File: src/components/ToggleButton/ToggleButton.tsx [hunk 3]
```diff
@@ -31,10 +46,26 @@ export function ToggleButton({ label, icon, onToggle, ...buttonProps }: Props) {
...
 }
+
+function sendOrder(
+    isOn: boolean,
+    id: number,
+    sendJsonMessage: SendJsonMessage
+) {
+    const controlOrder: ControlOrder = { id: id, state: isOn };
+    sendJsonMessage(controlOrder);
+    console.log("Control order sent!");
```
Comment (smorsua) [inline src/components/ToggleButton/ToggleButton.tsx]: Remove log

File: src/pages/TestingPage/TestControls/useControlForm.ts [hunk 4]
```diff
@@ -29,9 +30,10 @@ export function useControlForm(
         dispatch({ type: "RESET_INITIAL_STATE", payload: initialState });
     };
 
-    const SubmitHandler: SubmitHandler = () => {
+    const SubmitHandler: SubmitHandler = (sendJsonMessage: SendJsonMessage) => {
         if (form.isValid) {
-            //TODO: take the data, send FormData al backend? To be defined
+            sendJsonMessage(form.formData);
+            console.log("Form Data sent!");
```
Comment (smorsua) [inline src/pages/TestingPage/TestControls/useControlForm.ts]: Remove log

File: src/pages/TestingPage/TestingPage.tsx [hunk 5]
```diff
@@ -1,15 +1,78 @@
...
+        sendMessage,
+        sendJsonMessage,
+        lastMessage,
+        lastJsonMessage,
+        readyState,
+        getWebSocket,
+    } = useWebSocket(WEBSOCKET_URL, {
+        onOpen: () => console.log("Connection opened"),
+        shouldReconnect: (closeEvent) => {
+            console.log("Disconnected");
```
Comment (smorsua) [inline src/pages/TestingPage/TestingPage.tsx]: Remove log

File: src/pages/TestingPage/TestingPage.tsx [hunk 6]
```diff
@@ -1,15 +1,78 @@
...
+        if (lastJsonMessage !== null) {
+            const newVehicleState = lastJsonMessage as VehicleState;
+            if (newVehicleState.duty >= 0 && newVehicleState.duty < 256) {
+                setVehicleState(newVehicleState);
+            }
+        }
+    }, [lastJsonMessage]);
+
+    useEffect(() => {
+        console.log(vehicleState);
```
Comment (smorsua) [inline src/pages/TestingPage/TestingPage.tsx]: remove useEffect

**Review by smorsua — APPROVED** [review]
Looks good

**Review by jmaralo — APPROVED** [review]

**Pull request merged**